[6.3 🍒][CSOptimizer] Also favor existential metatype candidates in paramType isAny fast path

clackary opened the pull request:
- **Explanation**:
Resolves a regression in overload resolution with metatype candidates.

- **Scope**:
Besides resolving the regression, the fix adds an additional fast path for metatypes, so it is not expected to negatively impact performance.
- **Issues**:
Resolves: https://github.com/swiftlang/swift/issues/85020
Resolves: rdar://163074598
- **Original PRs**:
https://github.com/swiftlang/swift/pull/85652
- **Risk**:
Low. The change only tweaks overload scoring when the parameter is Any/Any.Type, returning to the same behavior as is in current/previous releases.
- **Testing**:
Adds a test case to catch this specific behavior. I also built a custom toolchain with this fix and verified that the issue is resolved in the original project in which it was discovered.
- **Reviewers**:
@xedin

clackary committed:
[6.3][CSOptimizer] test: Add test case for issue #85020

Overload resolution picks wrong initializer for some metatype arguments

https://github.com/swiftlang/swift/issues/85020

clackary committed:
[6.3][CSOptimizer] favor Any.Type for metatype args alongside Any fast-path

Resolves https://github.com/swiftlang/swift/issues/85020

xedin reviewed (approved):
Thank you again for the report and the fix!

xedin commented:
@swift-ci please test macOS platform

finagolfin committed:
[Test] Mark new `Interpreter/issue-85020` test as executable_test

Otherwise, various [non-executable CI like Android fail](https://ci.swift.org/job/oss-swift-package-swift-sdk-for-android/110/console).

(cherry picked from commit f1e77b776c0e59f152845007752aa7532f0394e8)

xedin commented:
@swift-ci please test
